File: src/com/GlehuDonsa/Blueprints/Spell.java
package com.GlehuDonsa.Blueprints;

public interface Spell
{
	String SPELL_NAME	= "";
	String CAST_TYPE 	= "";	// TRIGGER, TOGGLE, CHANNEL, PASSIVE
	String CAST_TARGET 	= "";	// UNIT, SELF, AUTO, GROUND, DIRECTION
	String SPELL_TYPE 	= "";	// DAMAGE, HEALING, MOVEMENT
	String SPELL_TARGET = "";	// SINGLE, AREA


	int castRange		= 0;
	int baseDamage 		= 0;
	int upgradeDamage 	= 0;
	int upgradeCooldown	= 0;
	int upgradeMax 		= 0;	// Max level

	int currentDmg		= 0;

	int manacost		= 0;
	int level			= 0;
	int baseCooldown	= 0;	// Frames

	int currentCd		= 0;

	boolean isExpired	= false;

	boolean canHitTarget(Champ target);
	void onHit(Champ target);

	void cooldown();

	void upgrade();

	String getInfo();

	boolean isExpired();
}

File: core/src/com/glehudonsa/battlearena/blueprints/Spell.kt
package com.glehudonsa.battlearena.blueprints

interface Spell
{

	enum class CastType {
		TRIGGER, TOGGLE, CHANNEL, PASSIVE
	}

	enum class CastTarget {
		UNIT, SELF, AUTO, GROUND, DIRECTION, ONHIT, NON
	}

	enum class SpellType {
		DAMAGE, HEALING, MOVEMENT
	}

	enum class SpellTarget {
		SINGLE, AREA
	}

	/* Kotlin is useful here because in Java you would need
	 *
	 * String getSpellName()
	 *   and
	 * @Override public String getSpellName() { return … };
	 *
	 * everytime. Kotlin is much shorter*/

	val spellName:		String
	val castType:		CastType
	val castTarget:		CastTarget
	val spellType:		SpellType
	val spellTarget:	SpellTarget

	val castRange:		Int
	val baseDamage:		Int
	val upgradeDamage:	Int
	val upgradeCooldown:Int
	val upgradeMax:		Int // Max level

	var manaCost:		Int
	var level:			Int
	val baseCooldown:	Int // Frames

	fun canHitTarget(target: Champ) = target.canBeHitSpell()
	fun onHit(target: Champ)

	fun cooldown()

	fun upgrade()
	{
		if(level < upgradeMax)
		{
			level++;
		}
	}

	fun isExpired(): Boolean

	fun getInfo() =
"""--------------------------------------------------------------------------
$spellName
Cast type:		$spellName
Cast target:	$castTarget
Spell type:		$spellType
Spell target:	$spellTarget
Description:
$desc
--------------------------------------------------------------------------"""

	val desc:		String

	// A function. Can be called from java: getCurrentDmg()
	val currentDmg
		get() = baseDamage + (level * upgradeDamage)

	val currentCd
		get() = baseCooldown - (level * upgradeCooldown)

}
